### PR TITLE
DPL: streamline DataAllocator::make

### PR DESCRIPTION
**Pull request by ktf**

### CODE REVIEW

**Comment by ktf** [comment]
@matthiasrichter this is how I would streamline the `DataAllocator::make` API using C++17 `if constexpr` in place of all the `std::enable_if`. It would also have the advantage that we can probably separate the various implementations (boost, arrow, ROOT) in different headerfiles with a bit more gymnastic.

This is WIP because there is something which needs to change in order to be able to use `auto&` everywhere (right now it would segfault for one of the cases) and because I think it can be simplifies even further.

**Comment by matthiasrichter** [comment]
@ktf, think we should push this forward. I have rebased the branch and included the patch for running also the workflow unit tests (PR #2140).

What is needed to be changed to get this PR ready for merging?

**Comment by ktf** [comment]
As discussed privately, there is an issue with references, IIRC. We should resolve that before merging this.

**Comment by matthiasrichter** [comment]
I think we also see it in the CI, will try to have a look
```
.../Framework/Core/include/Framework/DataAllocator.h: In instantiation of 'auto& o2::framework::DataAllocator::make(const o2::framework::Output&, Args ...) [with T = arrow::ipc::RecordBatchWriter; Args = {std::shared_ptr<arrow::Schema>}]':
.../Framework/Core/src/AODReaderHelpers.cxx:209:110:   required from here
.../Framework/Core/include/Framework/DataAllocator.h:145:58: error: reference to local variable 'writer' returned [-Werror=return-local-addr]
           std::shared_ptr<arrow::ipc::RecordBatchWriter> writer;
                                                          ^~~~~~
```

**Comment by ktf** [comment]
yes, exactly.

File: Framework/Core/include/Framework/DataAllocator.h [hunk 1]
```diff
@@ -84,98 +99,64 @@ class DataAllocator
...
+  /// Generic helper to create an object which is owned by the framework and
+  /// returned as a reference to the own object.
+  /// Note: decltype(auto) will deduce the return type from the expression and it
+  /// will be lvalue reference for the framework-owned objects. Instances of local
+  /// variables like shared_ptr will be returned by value/move/return value optimization.
+  /// Objects created this way will be sent to the channel specified by @spec
   template <typename T, typename... Args>
-  typename std::enable_if<std::is_base_of<TableBuilder, T>::value == true, T&>::type
-    make(const Output& spec, Args... args)
+  decltype(auto) make(const Output& spec, Args... args)
```
Comment (matthiasrichter) [inline Framework/Core/include/Framework/DataAllocator.h]: @ktf that seems to do the trick

**Comment by ktf** [comment]
Look good to me. Thanks. I am merging this.

**Pull request merged**